presto-main/src/main/java/com/facebook/presto/sql/planner/GroupingOperationRewriter.java

```diff
@@ -0,0 +1,92 @@
+import com.facebook.presto.spi.type.Type;
+import com.facebook.presto.spi.type.TypeSignature;
+import com.facebook.presto.sql.analyzer.Analysis;
+import com.facebook.presto.sql.tree.Expression;
+import com.facebook.presto.sql.tree.ExpressionRewriter;
+import com.facebook.presto.sql.tree.ExpressionTreeRewriter;
+import com.facebook.presto.sql.tree.FunctionCall;
+import com.facebook.presto.sql.tree.GroupingOperation;
+import com.facebook.presto.sql.tree.QuerySpecification;
+import com.facebook.presto.type.ListLiteralType;
+import com.google.common.collect.ImmutableList;
+
+import java.util.Arrays;
+import java.util.IdentityHashMap;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import static com.facebook.presto.spi.type.BigintType.BIGINT;
+import static com.facebook.presto.spi.type.IntegerType.INTEGER;
+import static com.facebook.presto.sql.analyzer.ExpressionAnalyzer.resolveFunction;
+import static java.util.Objects.requireNonNull;
+
+public class GroupingOperationRewriter
+        extends ExpressionRewriter<Void>
+{
+    private final PlanBuilder subPlan;
+    private final QuerySpecification queryNode;
+    private final Analysis analysis;
+    private final Metadata metadata;
+    private final boolean isGroupingOperationTopLevel;
+
+    public GroupingOperationRewriter(PlanBuilder subPlan, QuerySpecification queryNode, Analysis analysis, Metadata metadata, boolean isGroupingOperationTopLevel)
+    {
+        this.subPlan = requireNonNull(subPlan, "subPlan is null");
+        this.queryNode = requireNonNull(queryNode, "node is null");
+        this.analysis = requireNonNull(analysis, "analysis is null");
+        this.metadata = requireNonNull(metadata, "metadata is null");
+        this.isGroupingOperationTopLevel = isGroupingOperationTopLevel;
+    }
+
+    @Override
+    public Expression rewriteExpression(Expression node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+    {
+        return treeRewriter.defaultRewrite(node, context);
+    }
+
+    @Override
+    public Expression rewriteGroupingOperation(GroupingOperation node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+    {
+        FunctionCall rewrittenExpression = subPlan.rewriteGroupingOperationToFunctionCall(node, queryNode);
+        if (isGroupingOperationTopLevel) {
+            ImmutableList.Builder<Expression> outputExpressionBuilder = ImmutableList.builder();
+            outputExpressionBuilder.addAll(analysis.getOutputExpressions(queryNode).stream()
+                    .map((outputExpression) -> (outputExpression instanceof GroupingOperation) ? rewrittenExpression : outputExpression)
+                    .collect(Collectors.toList()));
+            analysis.setOutputExpressions(node, outputExpressionBuilder.build());
+        }
+        IdentityHashMap<Expression, Type> expressionTypes = new IdentityHashMap<>();
+        IdentityHashMap<FunctionCall, Signature> functionSignatures = new IdentityHashMap<>();
```

Comment:
Good to know. Unfortunately I don't think there's much I can do now. I could convert these two variables and the two methods of `Analysis` to use `IdentityMap` but that wouldn't be complete and hacky. I say whichever patch gets merged second will have to rebase.

cc @ArturGajowy